api: initial gpio apis for review

posborne opened the pull request:
CC @nastevens for review.

posborne committed:
api: initial gpio apis for review

nastevens commented on src/lib.rs:
```diff
@@ -1,3 +1,70 @@
+/// Generic trait providing access to a single GPIO
+pub trait Gpio {
+    /// Get the currently configured direciton of this GPIO
+    fn get_direction(&mut self) -> Result<Direction>;
```
Does `self` need to be `mut` here? It seems like even if some sort of mutability ended up being needed, it could be done through a `Cell`.

posborne replied:
I haven't really used `Cell` or `RefCell` as of yet, but it does appear to allow for dealing with the case that I was most concerned with: https://doc.rust-lang.org/std/cell/#implementation-details-of-logically-immutable-methods.  As such, I agree that these should take `&self` and can do so without making it unreasonable for an implementer to make use of interior mutability if required.  The facade should be immutable.

nastevens commented on src/lib.rs:
```diff
@@ -1,3 +1,70 @@
+    /// Set the pin to be an output with the provided level
+    fn set_output(&mut self, value: Level) -> Result<()>;
+    /// Get the current level of this GPIO (input only)
+    fn get_level(&mut self) -> Result<Level>;
```
Same idea as above - not sure if `mut` is needed here.

nastevens commented:
Sorry it took me so long to look at this. It's always hard to guess ahead what APIs will be sufficient for all cases, but this seems like a strong start. :+1:

posborne committed:
api: make getters immutable references

As per code review, if mutability is required by
implementors it can be handled via internal
mutability via something like Cell/RefCell.

Signed-off-by: Paul Osborne <osbpau@gmail.com>

posborne committed:
api: add support for stable and nightly

By means of hacks!

Signed-off-by: Paul Osborne <osbpau@gmail.com>

cwoodall commented:
I recently on the Cortex-M processors came across some features of GPIOs that would be nice to access namely:
- open-drain configurations
- slew rate control
- pin change interrupts

I think as a base API this is wonderful though, I am not particularly sure such non-portable details should clutter this particular API

nastevens commented:
@cwoodall - open-drain/open-source configurations are pretty common across GPIOs. My opinion is that these could be added without clutter to the API.

Slew rate control is difficult because these are usually controlled by something vague like "fast" and "slow", or "high", "medium", "low". In this case an extended trait like `GpioSlewRate` would probably be best.

Pin change interrupts are super interesting - it'd be really cool to be able to fire a closure based on an interrupt! Unfortunately, they are even less portable than the slew rate settings, and depend not only on the GPIO hardware but on the interrupt hardware and OS. A design for an interrupt library would probably be needed first.
